service/src/BatchClient.cpp

```diff
@@ -0,0 +1,120 @@
+                                                          int num_control)
+    {
+        return geopm::make_unique<BatchClientImp>(server_key, timeout,
+                                                  num_signal, num_control);
+    }
+
+    BatchClientImp::BatchClientImp(const std::string &server_key, double timeout,
+                                   int num_signal, int num_control)
+        : BatchClientImp(num_signal, num_control,
+                         BatchStatus::make_unique_client(server_key),
+                         num_signal == 0 ? nullptr :
+                         SharedMemory::make_unique_user(BatchServer::M_SHMEM_PREFIX +
+                                                        server_key + "-signal",
+                                                        timeout),
+                         num_control == 0 ? nullptr :
+                         SharedMemory::make_unique_user(BatchServer::M_SHMEM_PREFIX +
+                                                        server_key + "-control",
+                                                        timeout))
+    {
+
+    }
+
+
+    BatchClientImp::BatchClientImp(int num_signal, int num_control,
+                                   std::shared_ptr<BatchStatus> batch_status,
+                                   std::shared_ptr<SharedMemory> signal_shmem,
+                                   std::shared_ptr<SharedMemory> control_shmem)
+        : m_num_signal(num_signal)
+        , m_num_control(num_control)
+        , m_batch_status(batch_status)
+        , m_signal_shmem(signal_shmem)
+        , m_control_shmem(control_shmem)
+    {
+
+    }
+
+    std::vector<double> BatchClientImp::read_batch(void)
+    {
+        if (m_num_signal == 0) {
+            return {};
+        }
+        m_batch_status->send_message(BatchStatus::M_MESSAGE_READ);
+        m_batch_status->receive_message(BatchStatus::M_MESSAGE_CONTINUE);
+        double *buffer = (double *)m_signal_shmem->pointer();
+        std::vector<double> result(buffer, buffer + m_num_signal);
+        return result;
+    }
+
+    void BatchClientImp::write_batch(std::vector<double> settings)
+    {
+        if (settings.size() != (size_t)m_num_control) {
+            throw Exception("BatchClientImp::write_batch(): settings vector length does not match the number of configured controls",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (m_num_control == 0) {
+            return;
+        }
+        double *buffer = (double *)m_control_shmem->pointer();
+        std::copy(settings.begin(), settings.end(), buffer);
+        m_batch_status->send_message(BatchStatus::M_MESSAGE_WRITE);
+        m_batch_status->receive_message(BatchStatus::M_MESSAGE_CONTINUE);
+    }
+
+    void BatchClientImp::stop_batch(void)
+    {
+        m_batch_status->send_message(BatchStatus::M_MESSAGE_QUIT);
+        m_batch_status->receive_message(BatchStatus::M_MESSAGE_QUIT);
```

Comment:
Will this result in left-over resources on the server side if a client terminates without calling stop_batch()?

I think you already answered this when reviewing the server side, but I can't remember. Maybe a comment here (or when creating this connection) could help.


Also why does this need to receive a quit message? Is there some subsequent work that fails if the server hasn't yet ack'd the stop request?

Reply:
The geopmd process checks that all PID's that have open sessions are active once per second.  If a PID that has an open session terminates, geopmd will stop any batch server running for that session and restore all control values.  This is done with the callback registration below:

https://github.com/geopm/geopm/blob/dev/service/geopmdpy/service.py#L680-L687

There is a problem though... the close_session() method does not shut down the open batch server as it should...

https://github.com/geopm/geopm/blob/dev/service/geopmdpy/service.py#L419-L428

Reply:
Addressed this issue with the rather large patch added to the next PR, #1998

https://github.com/geopm/geopm/pull/1998/commits/b797fba5e412953875aa8442e837251b7dd84c95

Since this does not impact the BatchClient implementation, this can be merged with the #1998

Reply:
Oh and the QUIT message. We send that message whenever the client has made a request and we are ending.  In the highlighted case, we send back the QUIT message when the client requested it.  If the server is interrupted by geopmd while servicing a request from the client, the client will wait for the CONTINUE message, but actually receive the QUIT message.  This will be an error on the client side *unless* the request that got interrupted was the QUIT request...